File: Matlab/ANCOVA.m
classdef ANCOVA

    methods(Static)

        function [choosen_landmarks]= analysis_ANCOVA (file_name,all_data,all_genders, all_ages, landmark_names, title_str)% Carica i tuoi dati
        
            fileID = fopen(file_name, 'w');
            % Inizializza la matrice per i risultati
            p = zeros(size(all_data, 1), 3);
            significant_landmark = zeros(size(all_data, 1), 3);
            F_Gender = zeros(size(all_data,1),1);
            F_Age= zeros(size(all_data,1),1);
            F_Interaction=zeros(size(all_data,1),1);
            choosen_landmarks = cell(3, 1);


            ANCOVA_Analysis = './ANCOVA_Analysis';
            if ~exist(ANCOVA_Analysis, 'dir')
                mkdir(ANCOVA_Analysis);
            end
            
            title_new = strcat ('ANCOVA analysis with age and gender analysis-',title_str);
            
           for i = 1:size(all_data, 1)
                % Crea una tabella con i tuoi dati
                t = table(all_data(i, :)', double(all_genders'), double(all_ages'), 'VariableNames', {'Landmark', 'Gender', 'Age'});
                
                % Adatta un modello ANCOVA ai tuoi dati
                mdl = fitlm(t, 'Landmark ~ Gender + Age + Gender:Age');
                
                % Ottieni i p-values dal modello
                p(i,:) = mdl.Coefficients.pValue(2:end)';
                
                % Ottieni le statistiche F dal modello
                F_Gender(i,:) = mdl.Coefficients.tStat(2);
                F_Age(i,:) = mdl.Coefficients.tStat(3);
                F_Interaction(i,:) = mdl.Coefficients.tStat(4);
            end

            
        
        
            for i=1:length(landmark_names)
                % A low p-value indicates that it is unlikely to obtain the observed data if the null hypothesis (i.e., there 
                % is no difference between the groups) were true.
        
                % The F ratio is the ratio between the variance between groups and the variance within groups. 
                % An F ratio of 1 suggests that the variance between groups is equal to the variance within groups. 
                % Therefore, an F ratio greater than 1 might suggest that there is a significant difference between the groups. 
                % A large F ratio with a low p-value (below the significance level) indicates that it is unlikely to obtain 
                % the observed data if the null hypothesis were true, suggesting that there is a significant difference between the groups.
                significant_landmark(i, :) = p(i,1:3) < 0.05 & ( F_Gender(i) > 1 | F_Age(i) >1 | F_Interaction(i) >1);
                
            end
        
            
            % A significant interaction effect between gender and age would suggest that the effect of gender on landmarks depends on age,
            % or vice versa. For example, it could be that a particular landmark differs significantly between men and women, but only in 
            % a particular age group.
        

            significant_gender_indices = find(significant_landmark(:, 1) == 1);
            
            significant_age_indices = find(significant_landmark(:, 2) == 1);
            
            significant_interaction_indices = find(significant_landmark(:, 3) == 1);
            
        
            fprintf(fileID,'ANCOVA results %s\n', title_str);
            
        
            if (~isempty(significant_gender_indices)) 
                choosen_landmarks{1} = significant_gender_indices;
                fprintf(fileID,'\nIt allows me to say that for the gender analysis, the most significant landmark are:\n');
                for i=1:length(significant_gender_indices)
                    fprintf(fileID,'%s\n', landmark_names{significant_gender_indices(i)});
                end
            else
                fprintf(fileID, ...
                    '\nIt allows me to say that for the gender analysis \n there are no significant landmark for classification by Gender\n');
            end
            if (~isempty(significant_age_indices)) 
                choosen_landmarks{2} = significant_age_indices;
                fprintf(fileID,'\nIt allows me to say that for the age analysis, the most significant landmark are:\n\n');
                for i=1:length(significant_age_indices)
                    fprintf(fileID,'%s\n', landmark_names{significant_age_indices(i)});
                end
            else
                fprintf(fileID,['\nIt allows me to say that for the gender analysis\n' ...
                    'there are no significant landmark for classification by Age\n\n']);
            end
            if (~isempty(significant_interaction_indices)) 
                choosen_landmarks{3} = significant_interaction_indices;
                fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and gender, \n' ...
                    'the most significant landmark are:\n\n']);
                
                for i=1:length(significant_interaction_indices)
                    fprintf(fileID,'%s\n', landmark_names{significant_interaction_indices(i)});
                end
            else
               fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and gender \n' ...
                    'there are no significant landmark for classification by Age and Gender\n\n']);
            end
        
            f1=figure;
            bar(p);
            xticks(1:size(all_data, 1));
            xticklabels(landmark_names);
            xtickangle(45);
            xlabel('Landmark');
            ylabel('p-value');
            title(title_new);
            
            % Aggiungi una linea orizzontale al livello di significatività
            hold on;
            line(xlim, [0.05, 0.05], 'Color', 'red', 'LineStyle', '--');
            hold off;
            
            % Mostra la legenda
            legend({'Gender', 'Age', 'Interaction'}, 'Location', 'northoutside');

            
            title_new = strrep(title_str,' ','_');
            name_file = sprintf('ANCOVA_%s',title_new);
            name_file = strcat(name_file, '.png');
            saveas(f1, fullfile('ANCOVA_Analysis', name_file)); 
        
        fclose(fileID);
                
        end




        %%%%%%%%%%%%%%%%%%%% for BMI analysis 

        function [choosen_landmarks]= analysis_ANCOVA_bmi (name_file,all_data,all_genders, all_ages, all_bmi, landmark_names, title_str)% Carica i tuoi dati


                ANCOVA_Analysis = './ANCOVA_Analysis';
                if ~exist(ANCOVA_Analysis, 'dir')
                    mkdir(ANCOVA_Analysis);
                end
                % Inizializza la matrice per i risultati
                p = zeros(size(all_data, 1), 7);
                significant_landmark = zeros(size(all_data, 1), 7);
                F_Gender = zeros(size(all_data,1),1);
                F_Age= zeros(size(all_data,1),1);
                F_BMI = zeros(size(all_data,1),1);
                F_Interaction_GA=zeros(size(all_data,1),1);
                F_Interaction_GB=zeros(size(all_data,1),1);
                F_Interaction_AB=zeros(size(all_data,1),1);
                F_Interaction_GAB=zeros(size(all_data,1),1);
                choosen_landmarks = cell(7, 1);
                title_new = strcat ('ANCOVA analysis with BMI factor-',title_str);
                fileID = fopen(name_file,'w');

                fprintf(fileID,'ANCOVA analysis with BMI factor %s\n', title_str);
                
                for i = 1:size(all_data, 1)
                    % Crea una tabella con i tuoi dati
                    t = table(all_data(i, :)', double(all_genders'), double(all_ages'), all_bmi', 'VariableNames', {'Landmark', 'Gender', 'Age', 'BMI'});
                    
                    % Adatta un modello ANCOVA ai tuoi dati
                    mdl = fitlm(t, 'Landmark ~ Gender + Age + BMI + Gender:Age + Gender:BMI + Age:BMI + Gender:Age:BMI');
                    
                    % Ottieni i p-values dal modello
                    p(i,:) = mdl.Coefficients.pValue(2:end)';
                    
                    % Ottieni le statistiche F dal modello
                    F_Gender(i,:) = mdl.Coefficients.tStat(2);
                    F_Age(i,:) = mdl.Coefficients.tStat(3);
                    F_BMI(i,:) = mdl.Coefficients.tStat(4);
                    F_Interaction_GA(i,:) = mdl.Coefficients.tStat(5);
                    F_Interaction_GB(i,:) = mdl.Coefficients.tStat(6);
                    F_Interaction_AB(i,:) = mdl.Coefficients.tStat(7);
                    F_Interaction_GAB(i,:) = mdl.Coefficients.tStat(8);
                    
                end

                 for i=1:length(landmark_names)
                   

                    % - p(i,1) is the p-value for the effect of gender. A low p-value suggests that gender has a significant effect 
                    % on the response variable.
                    % - p(i,2) is the p-value for the effect of age. A low p-value suggests that age has a significant effect on 
                    % the response variable.
                    % - p(i,3) is the p-value for the effect of BMI. A low p-value suggests that BMI has a significant effect on the 
                    % response variable.
                    % - p(i,4) is the p-value for the interaction between gender and age. A low p-value suggests that the effect of 
                    % gender on the response variable depends on age.
                    % - p(i,5) is the p-value for the interaction between gender and BMI. A low p-value suggests that the effect of 
                    % gender on the response variable depends on BMI.
                    % - p(i,6) is the p-value for the interaction between age and BMI. A low p-value suggests that the effect of age 
                    % on the response variable depends on BMI.
                    % - p(i,7) is the p-value for the interaction between gender, age, and BMI. A low p-value suggests that the effect 
                    % of gender on the response variable depends on both age and BMI.
                    
                    significant_landmark(i, :) = p(i,1:7) < 0.05 & ( F_Gender(i) > 1 | F_Age(i) >1 | F_BMI(i)>1| F_Interaction_GA(i) >1 | F_Interaction_GB(i)>1 | F_Interaction_AB(i)>1 | F_Interaction_GAB(i)>1);
                 
                 end

                significant_gender_indices = find(significant_landmark(:, 1) == 1);
                
                significant_age_indices = find(significant_landmark(:, 2) == 1);
                
                significant_bmi_indices = find(significant_landmark(:, 3) == 1);
            
                significant_interactionGA_indices = find(significant_landmark(:, 4) == 1);
            
                significant_interactionGB_indices = find(significant_landmark(:, 5) == 1);
            
                significant_interactionAB_indices = find(significant_landmark(:, 6) == 1);
            
                significant_interactionGAB_indices = find(significant_landmark(:,7) == 1);



                if (~isempty(significant_gender_indices)) 
                    choosen_landmarks{1} = significant_gender_indices;
                    fprintf(fileID,'\nIt allows me to say that for the gender analysis, the most significant landmark are:\n');
                    for i=1:length(significant_gender_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_gender_indices(i)});
                    end
                else
                    fprintf(fileID,['\nIt allows me to say that for the gender analysis \n' ...
                        'there are no significant landmark for classification by Gender\n']);
                end
                if (~isempty(significant_age_indices)) 
                    choosen_landmarks{2} = significant_age_indices;
                    fprintf(fileID,'\nIt allows me to say that for the age analysis, the most significant landmark are:\n');
                    for i=1:length(significant_age_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_age_indices(i)});
                    end
                else
                    fprintf(fileID,['\nIt allows me to say that for the gender analysis\n' ...
                        'there are no significant landmark for classification by Age\n\n']);
                end
                if (~isempty(significant_bmi_indices)) 
                    choosen_landmarks{3} = significant_bmi_indices;
                    fprintf(fileID,'\nIt allows me to say that for the BMI analysis, the most significant landmark are:\n');
                    for i=1:length(significant_bmi_indices)
                        fprintf(fileID,'%s\n\n', landmark_names{significant_bmi_indices(i)});
                    end
                else
                    fprintf(fileID,['\nIt allows me to say that for the BMI analysis\n' ...
                        'there are no significant landmark for classification by BMI \n\n']);
                end
            
                if (~isempty(significant_interactionGA_indices)) 
                    choosen_landmarks{4} = significant_interactionGA_indices;
                    fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and gender, \n' ...
                        'the most significant landmark are:\n\n']);
                    
                    for i=1:length(significant_interactionGA_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_interactionGA_indices(i)});
                    end
                else
                   fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and gender \n' ...
                        'there are no significant landmark for classification by Age and Gender\n\n']);
                end
            
            
                 if (~isempty(significant_interactionGB_indices)) 
                    choosen_landmarks{5} = significant_interactionGB_indices;
                    fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of gender and BMI, \n' ...
                        'the most significant landmark are:\n\n']);
                    
                    for i=1:length(significant_interactionGB_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_interactionGB_indices(i)});
                    end
                else
                   fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of  gender and BMI \n' ...
                        'there are no significant landmark for classification by Gender and BMI \n\n']);
                 end
            
            
                 if (~isempty(significant_interactionAB_indices)) 
                    choosen_landmarks{6} = significant_interactionAB_indices;
                    fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and BMI, \n' ...
                        'the most significant landmark are:\n\n']);
                    
                    for i=1:length(significant_interactionAB_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_interactionAB_indices(i)});
                    end
                else
                   fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age and BMI \n' ...
                        'there are no significant landmark for classification by age and BMI \n\n']);
                 end
            
                 if (~isempty(significant_interactionGAB_indices)) 
                    choosen_landmarks{7} = significant_interactionGAB_indices;
                    fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age, gender and BMI, \n' ...
                        'the most significant landmark are:\n\n']);
                    
                    for i=1:length(significant_interactionGAB_indices)
                        fprintf(fileID,'%s\n', landmark_names{significant_interactionGAB_indices(i)});
                    end
                else
                   fprintf(fileID,['\nIt allows me to say that for the analysis of the interaction of age, gender and BMI \n' ...
                        'there are no significant landmark for classification by age, gender and BMI \n\n']);
                end
            
                f2=figure;
                bar(p);
                xticks(1:size(all_data, 1));
                xticklabels(landmark_names);
                xtickangle(45);
                xlabel('Landmark');
                ylabel('p-value');
                title(title_new);
                
                % Aggiungi una linea orizzontale al livello di significatività
                hold on;
                line(xlim, [0.05, 0.05], 'Color', 'red', 'LineStyle', '--');
                hold off;
                
                % Mostra la legenda
                legend({'Gender', 'Age', 'BMI', 'Gender-Age', 'Gender-BMI', 'Age-BMI', 'Gender-Age-BMI'}, 'Location', 'northoutside');
            
                
                title_new = strrep(title_str,' ','_');
                name_file = sprintf('ANCOVA_BMI_%s',title_new);
                name_file = strcat(name_file, '.png');
                saveas(f2, fullfile('ANCOVA_Analysis', name_file)); % 
                fclose(fileID);
                
            end


    end
end

File: Python/ANCOVA.py
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.formula.api import ols

class ANCOVA:
    
    @staticmethod
    def analysis_ANCOVA(file_name, all_data, all_genders, all_ages, landmark_names, title_str):
        with open(file_name, 'w') as fileID:
            p = np.zeros((all_data.shape[0], 3))
            significant_landmark = np.zeros((all_data.shape[0], 3))
            F_Gender = np.zeros((all_data.shape[0], 1))
            F_Age = np.zeros((all_data.shape[0], 1))
            F_Interaction = np.zeros((all_data.shape[0], 1))
            chosen_landmarks = [None, None, None]

            if not os.path.exists('ANCOVA_Analysis'):
                os.makedirs('ANCOVA_Analysis')

            title_new = f'ANCOVA analysis with age and gender analysis-{title_str}'
            
            for i in range(all_data.shape[0]):
                df = pd.DataFrame({
                    'Landmark': all_data[i, :],
                    'Gender': all_genders,
                    'Age': all_ages
                })
                model = ols('Landmark ~ C(Gender) + Age + C(Gender):Age', data=df).fit()
                anova_table = sm.stats.anova_lm(model, typ=2)
                p[i, 0] = anova_table['PR(>F)']['C(Gender)']
                p[i, 1] = anova_table['PR(>F)']['Age']
                p[i, 2] = anova_table['PR(>F)']['C(Gender):Age']
                F_Gender[i] = anova_table['F']['C(Gender)']
                F_Age[i] = anova_table['F']['Age']
                F_Interaction[i] = anova_table['F']['C(Gender):Age']

            for i in range(len(landmark_names)):
                significant_landmark[i, 0] = (p[i, 0] < 0.05) and (F_Gender[i] > 1)
                significant_landmark[i, 1] = (p[i, 1] < 0.05) and (F_Age[i] > 1)
                significant_landmark[i, 2] = (p[i, 2] < 0.05) and (F_Interaction[i] > 1)

            significant_gender_indices = np.where(significant_landmark[:, 0] == 1)[0]
            significant_age_indices = np.where(significant_landmark[:, 1] == 1)[0]
            significant_interaction_indices = np.where(significant_landmark[:, 2] == 1)[0]



            fileID.write(f'ANCOVA results {title_str}\n')

            if len(significant_gender_indices) > 0:
                chosen_landmarks[0] = significant_gender_indices
                fileID.write('\nIt allows me to say that for the gender analysis, the most significant landmarks are:\n')
                for i in significant_gender_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the gender analysis\n there are no significant landmarks for classification by Gender\n')

            if len(significant_age_indices) > 0:
                chosen_landmarks[1] = significant_age_indices
                fileID.write('\nIt allows me to say that for the age analysis, the most significant landmarks are:\n\n')
                for i in significant_age_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the age analysis\n there are no significant landmarks for classification by Age\n\n')

            if len(significant_interaction_indices) > 0:
                chosen_landmarks[2] = significant_interaction_indices
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and gender,\n the most significant landmarks are:\n\n')
                for i in significant_interaction_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and gender\n there are no significant landmarks for classification by Age and Gender\n\n')

            plt.figure()
            bar_width = 0.25
            x = np.arange(p.shape[0])

            plt.bar(x, p[:, 0], bar_width, label='Gender')
            plt.bar(x + bar_width, p[:, 1], bar_width, label='Age')
            plt.bar(x + 2 * bar_width, p[:, 2], bar_width, label='Interaction')

            plt.xticks(x + bar_width, landmark_names, rotation=45)
            plt.xlabel('Landmark')
            plt.ylabel('p-value')
            plt.title(title_new)
            plt.axhline(y=0.05, color='r', linestyle='--')
            plt.legend(loc='upper right')
            plt.tight_layout()
            plt.savefig(f'ANCOVA_Analysis/ANCOVA_{title_str.replace(" ", "_")}.png')
            plt.close()


        return chosen_landmarks

    @staticmethod
    def analysis_ANCOVA_bmi(name_file, all_data, all_genders, all_ages, all_bmi, landmark_names, title_str):
        if not os.path.exists('ANCOVA_Analysis'):
            os.makedirs('ANCOVA_Analysis')
        
        p = np.zeros((all_data.shape[0], 7))
        significant_landmark = np.zeros((all_data.shape[0], 7))
        F_Gender = np.zeros((all_data.shape[0], 1))
        F_Age = np.zeros((all_data.shape[0], 1))
        F_BMI = np.zeros((all_data.shape[0], 1))
        F_Interaction_GA = np.zeros((all_data.shape[0], 1))
        F_Interaction_GB = np.zeros((all_data.shape[0], 1))
        F_Interaction_AB = np.zeros((all_data.shape[0], 1))
        F_Interaction_GAB = np.zeros((all_data.shape[0], 1))
        chosen_landmarks = [None] * 7
        title_new = f'ANCOVA analysis with BMI factor-{title_str}'

        with open(name_file, 'w') as fileID:
            fileID.write(f'ANCOVA analysis with BMI factor {title_str}\n')
            for i in range(all_data.shape[0]):
                df = pd.DataFrame({
                    'Landmark': all_data[i, :],
                    'Gender': all_genders,
                    'Age': all_ages,
                    'BMI': all_bmi
                })
                model = ols('Landmark ~ C(Gender) + Age + BMI + C(Gender):Age + C(Gender):BMI + Age:BMI + C(Gender):Age:BMI', data=df).fit()
                anova_table = sm.stats.anova_lm(model, typ=2)
                p[i, 0] = anova_table['PR(>F)']['C(Gender)']
                p[i, 1] = anova_table['PR(>F)']['Age']
                p[i, 2] = anova_table['PR(>F)']['BMI']
                p[i, 3] = anova_table['PR(>F)']['C(Gender):Age']
                p[i, 4] = anova_table['PR(>F)']['C(Gender):BMI']
                p[i, 5] = anova_table['PR(>F)']['Age:BMI']
                p[i, 6] = anova_table['PR(>F)']['C(Gender):Age:BMI']
                F_Gender[i] = anova_table['F']['C(Gender)']
                F_Age[i] = anova_table['F']['Age']
                F_BMI[i] = anova_table['F']['BMI']
                F_Interaction_GA[i] = anova_table['F']['C(Gender):Age']
                F_Interaction_GB[i] = anova_table['F']['C(Gender):BMI']
                F_Interaction_AB[i] = anova_table['F']['Age:BMI']
                F_Interaction_GAB[i] = anova_table['F']['C(Gender):Age:BMI']

            for i in range(len(landmark_names)):
                significant_landmark[i, 0] = (p[i, 0] < 0.05) and (F_Gender[i] > 1)
                significant_landmark[i, 1] = (p[i, 1] < 0.05) and (F_Age[i] > 1)
                significant_landmark[i, 2] = (p[i, 2] < 0.05) and (F_BMI[i] > 1)
                significant_landmark[i, 3] = (p[i, 3] < 0.05) and (F_Interaction_GA[i] > 1)
                significant_landmark[i, 4] = (p[i, 4] < 0.05) and (F_Interaction_GB[i] > 1)
                significant_landmark[i, 5] = (p[i, 5] < 0.05) and (F_Interaction_AB[i] > 1)
                significant_landmark[i, 6] = (p[i, 6] < 0.05) and (F_Interaction_GAB[i] > 1)

            significant_gender_indices = np.where(significant_landmark[:, 0] == 1)[0]
            significant_age_indices = np.where(significant_landmark[:, 1] == 1)[0]
            significant_bmi_indices = np.where(significant_landmark[:, 2] == 1)[0]
            significant_interactionGA_indices = np.where(significant_landmark[:, 3] == 1)[0]
            significant_interactionGB_indices = np.where(significant_landmark[:, 4] == 1)[0]
            significant_interactionAB_indices = np.where(significant_landmark[:, 5] == 1)[0]
            significant_interactionGAB_indices = np.where(significant_landmark[:, 6] == 1)[0]


            if len(significant_gender_indices) > 0:
                chosen_landmarks[0] = significant_gender_indices
                fileID.write('\nIt allows me to say that for the gender analysis, the most significant landmarks are:\n')
                for i in significant_gender_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the gender analysis\n there are no significant landmarks for classification by Gender\n')

            if len(significant_age_indices) > 0:
                chosen_landmarks[1] = significant_age_indices
                fileID.write('\nIt allows me to say that for the age analysis, the most significant landmarks are:\n')
                for i in significant_age_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the age analysis\n there are no significant landmarks for classification by Age\n\n')

            if len(significant_bmi_indices) > 0:
                chosen_landmarks[2] = significant_bmi_indices
                fileID.write('\nIt allows me to say that for the BMI analysis, the most significant landmarks are:\n')
                for i in significant_bmi_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the BMI analysis\n there are no significant landmarks for classification by BMI\n\n')

            if len(significant_interactionGA_indices) > 0:
                chosen_landmarks[3] = significant_interactionGA_indices
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and gender,\n the most significant landmarks are:\n\n')
                for i in significant_interactionGA_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and gender\n there are no significant landmarks for classification by Age and Gender\n\n')

            if len(significant_interactionGB_indices) > 0:
                chosen_landmarks[4] = significant_interactionGB_indices
                fileID.write('\nIt allows me to say that for the analysis of the interaction of gender and BMI,\n the most significant landmarks are:\n\n')
                for i in significant_interactionGB_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the analysis of the interaction of gender and BMI\n there are no significant landmarks for classification by Gender and BMI\n\n')

            if len(significant_interactionAB_indices) > 0:
                chosen_landmarks[5] = significant_interactionAB_indices
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and BMI,\n the most significant landmarks are:\n\n')
                for i in significant_interactionAB_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the analysis of the interaction of age and BMI\n there are no significant landmarks for classification by Age and BMI\n\n')

            if len(significant_interactionGAB_indices) > 0:
                chosen_landmarks[6] = significant_interactionGAB_indices
                fileID.write('\nIt allows me to say that for the analysis of the interaction of gender, age and BMI,\n the most significant landmarks are:\n\n')
                for i in significant_interactionGAB_indices:
                    fileID.write(f'{landmark_names[i]}\n')
            else:
                fileID.write('\nIt allows me to say that for the analysis of the interaction of gender, age and BMI\n there are no significant landmarks for classification by Gender, Age and BMI\n\n')

            plt.figure()
            bar_width = 0.1
            x = np.arange(p.shape[0])

            plt.bar(x, p[:, 0], bar_width, label='Gender')
            plt.bar(x + bar_width, p[:, 1], bar_width, label='Age')
            plt.bar(x + 2 * bar_width, p[:, 2], bar_width, label='BMI')
            plt.bar(x + 3 * bar_width, p[:, 3], bar_width, label='Interaction GA')
            plt.bar(x + 4 * bar_width, p[:, 4], bar_width, label='Interaction GB')
            plt.bar(x + 5 * bar_width, p[:, 5], bar_width, label='Interaction AB')
            plt.bar(x + 6 * bar_width, p[:, 6], bar_width, label='Interaction GAB')

            plt.xticks(x + 3 * bar_width, landmark_names, rotation=45)
            plt.xlabel('Landmark')
            plt.ylabel('p-value')
            plt.title(title_new)
            plt.axhline(y=0.05, color='r', linestyle='--')
            plt.legend(loc='upper right')
            plt.tight_layout()
            plt.savefig(f'ANCOVA_Analysis/ANCOVA_BMI_{title_str.replace(" ", "_")}.png')
            plt.close()


        return chosen_landmarks
